File: hw2/src/container.cpp
#include <iostream>
#include "container.h"
#include "input_helper.h"
#include "random.h"

Container::Container()
    : m_len(0), m_data{}
{
}

Container::~Container()
{
    for (size_t i = 0; i < m_len; i++)
    {
        delete m_data[i];
    }
}

void Container::In(std::istream& in)
{
    // проверка на пустой файл
    if (in.peek() == std::istream::traits_type::eof())
    {
        return;
    }

    while (!in.eof())
    {
        m_data[m_len++] = BaseTransport::Create(in);
    }
}

void Container::InRnd(std::istream& in)
{
    size_t count;
    int seed;
    if (!(std::cin >> count >> seed))
    {
        InputHelper::InvalidInput();
    }

    if (count < 0 || count > k_maxLength)
    {
        InputHelper::InvalidInput();
    }

    m_len = count;
    Random rng(seed);
    for (size_t i = 0; i < count; i++)
    {
        m_data[i] = BaseTransport::CreateRnd(in, rng);
    }
}

void Container::Out(std::ostream& out)
{
    for (size_t i = 0; i < m_len; i++)
    {
        m_data[i]->Out(out);
    }
}

void Container::Sort()
{
    BaseTransport* key;
    ptrdiff_t i, j;
    for (i = 1; i < static_cast<ptrdiff_t>(m_len); i++)
    {
        key = m_data[i];
        j = i - 1;

        while (j >= 0 && m_data[j]->GetIdealTime() > key->GetIdealTime())
        {
            m_data[j + 1] = m_data[j];
            j = j - 1;
        }

        m_data[j + 1] = key;
    }
}

File: hw3/src/container.py
'''
container.py - содержит тип данных, представляющий простейший контейнер
'''

import input_helper
import random
from base_transport import *

# Простейший контейнер на основе одномерного списка
class Container:
    MAX_LENGTH = 10000

    # Инициализация контейнера
    def __init__(self):
        self.len = 0
        self.data = [None] * self.MAX_LENGTH

    # Ввод содержимого контейнера
    def inp(self, file):
        for line in file:
            self.data[self.len] = BaseTransport.create(line)
            self.len += 1

    # Случайный ввод содержимого контейнера
    def inp_rnd(self, file):
        try:
            splitted = file.readline().split()
            count = int(splitted[0])
            seed = int(splitted[0])
            if count < 0 or count > self.MAX_LENGTH:
                raise Exception()
            random.seed(seed)
            self.len = count
            for i in range(count):
                self.data[i] = BaseTransport.create_rnd()
        except Exception:
            input_helper.invalid_input()

    # Вывод в список строк
    def out(self):
        return [str(self.data[i]) for i in range(self.len)]

    # Сортировка контейнера методом Straight Insertion
    def sort(self):
        for i in range(1, self.len):
            key = self.data[i]
            j = i - 1
            while j >= 0 and self.data[j].get_ideal_time() > key.get_ideal_time():
                self.data[j + 1] = self.data[j]
                j = j - 1
            self.data[j + 1] = key
